appserver/payara-appserver-modules/yubikey-authentication/src/main/java/fish/payara/security/annotations/TwoFactorAuthenticationMechanismDefinition.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+ * 
+ *    Copyright (c) [2018] Payara Foundation and/or its affiliates. All rights reserved.
+ * 
+ *     The contents of this file are subject to the terms of either the GNU
+ *     General Public License Version 2 only ("GPL") or the Common Development
+ *     and Distribution License("CDDL") (collectively, the "License").  You
+ *     may not use this file except in compliance with the License.  You can
+ *     obtain a copy of the License at
+ *     https://github.com/payara/Payara/blob/master/LICENSE.txt
+ *     See the License for the specific
+ *     language governing permissions and limitations under the License.
+ * 
+ *     When distributing the software, include this License Header Notice in each
+ *     file and include the License file at glassfish/legal/LICENSE.txt.
+ * 
+ *     GPL Classpath Exception:
+ *     The Payara Foundation designates this particular file as subject to the "Classpath"
+ *     exception as provided by the Payara Foundation in the GPL Version 2 section of the License
+ *     file that accompanied this code.
+ * 
+ *     Modifications:
+ *     If applicable, add the following below the License Header, with the fields
+ *     enclosed by brackets [] replaced by your own identifying information:
+ *     "Portions Copyright [year] [name of copyright owner]"
+ * 
+ *     Contributor(s):
+ *     If you wish your version of this file to be governed by only the CDDL or
+ *     only the GPL Version 2, indicate your decision by adding "[Contributor]
+ *     elects to include this software in this distribution under the [CDDL or GPL
+ *     Version 2] license."  If you don't indicate a single choice of license, a
+ *     recipient has the option to distribute your version of this file under
+ *     either the CDDL, the GPL Version 2 or to extend the choice of license to
+ *     its licensees as provided above.  However, if you add GPL Version 2 code
+ *     and therefore, elected the GPL Version 2 license, then the option applies
+ *     only if the new code is made subject to such option by the copyright
+ *     holder.
+ */
+
+package fish.payara.security.annotations;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import javax.enterprise.util.Nonbinding;
+import javax.security.enterprise.authentication.mechanism.http.LoginToContinue;
+
+/**
+ * Used to define a Two Factor Authentication Mechanism with an annotation and provide configuration options
+ * @author Mark Wareham
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+public @interface TwoFactorAuthenticationMechanismDefinition {
```

Comment:
I think this would be better named something more like YubikeyMechanismDefiniton, as a 2FA definition won't necessarily use Yubikey as one of its methods. If you do stick with TwoFactorAuthenticationMechanismDefinition, then using yubikey should be another option inside it.

Reply:
The 2FA Mechanism is completely independent of yubikey. You could use the 2FA mech with an LDAPIdentityStore and a DatabaseIdentityStore for example. It selects the identity stores based on what's defined in the app and their priority

Reply:
This is still not giving the option of multiple authentication mechanism, which is what 2FA is, instead this is making multiple identity stores used in an AND system rather than the default OR.

Reply:
I agree, it does not allow two authentication mechanisms, but two identity stores. I will rename to avoid confusion